File: python_implementation/sort.py
import numpy as np
import hungarian
import intersection
import ddekf
from scipy import linalg 
import math
from copy import copy, deepcopy


class trackedObject:
    def __init__(self,position,width,height,velocity,velocity_covariance,time,name,objectType = None ):
        self.position = position
        self.velocity = velocity
        self.velocity_covariance = velocity_covariance
        self.velocity_covariance_sqrt = np.linalg.cholesky(velocity_covariance).transpose()
        self.last_update = time
        self.name = name
        self.objectType = objectType
        self.last_predict = time
        self.width = width
        self.height = height


class measurement():
    def __init__(self,x,y,width,height,objectType = None,percentage=None):
        self.x = x
        self.y = y
        self.position = np.array([x,y])
        self.percentage = percentage
        self.objectType = objectType
        self.width = width
        self.height = height

class tracker:
   
    def __init__(self,sensorNoise=np.array([[1,0],[0,1]]),processNoise=np.array([[1,0,0,0],[0,1,0,0],[0,0,1,0],[0,0,0,1]]),predictFunction=None,predictFunctionJacobian=None,minimumIOU=0.001,newShapeWeight=1.0/10.0,threshold=4,initialVelocity=np.array([[0,0]]).transpose(),initialCovariance=np.array([[1,0,0,0],[0,1,0,0],[0,0,1,0],[0,0,0,1]])):

        self.trackedObjectsByType = {}
        self.processNoise = processNoise
        self.sensorNoise = sensorNoise
        self.processNoiseSqrt = np.linalg.cholesky(self.processNoise).transpose()
        self.sensorNoiseSqrt = np.linalg.cholesky(self.sensorNoise).transpose()
        self.measurementMatrix = np.array([[1,0,0,0],[0,1,0,0]])
        self.minimum_IOU = minimumIOU
        self.threshold = threshold
        self.candidateCount = 0
        self.newShapeWeight = newShapeWeight

        if predictFunction == None:
            self.predictFunction = self.predictFunctionDefault
        else:
            self.predictFunction = predictFunction

        if predictFunctionJacobian == None:
            self.predictFunctionJacobian = self.predictFunctionJacobianDefault
        else:
            self.predictFunctionJacobian = predictFunctionJacobian

        self.initialCovariance = initialCovariance
        self.initialVelocity = initialVelocity

    def get_sensorNoise(self):
        return copy.copy(self.sensorNoise)

    def set_minimumIOU(self,iou):
        self.minimum_IOU = iou
        
    def set_threshold(self,threshold):
        self.threshold = threshold

    def set_sensorNoise(self,sensorNoise):
        self.sensorNoise = sensorNoise
        self.sensorNoiseSqrt = np.linalg.cholesky(self.sensorNoise).transpose()

    def get_processNoise(self):
        return copy.copy(self.processNoise)

    def set_processNoise(self,processNoise):
        self.processNoise = processNoise
        self.processNoiseSqrt = np.linalg.cholesky(self.processNoise).transpose()

    def set_predictFunction(self,function,jacobian):
        self.predictFunction = function
        self.predictFunctionJacobian = jacobian

    def set_newShapeWeight(self,weight):
        assert 0 <= weight and weight <= 1
        self.newShapeWeight = weight

    def predictFunctionDefault(self,state,t):
        position = state[0:2]
        velocity = state[2:4]
        position_next = position + velocity
        velocity_next = velocity
        result = [position_next[0], position_next[1], velocity_next[0], velocity_next[1]] 
        return np.array( result )

    def predictFunctionJacobianDefault(self,velocity,t):
        return np.array([[1,0,1,0],[0,1,0,1],[0,0,1,0],[0,0,0,1]]) 
    
    def getTracked(self):
        resultArray = []
        for key in self.trackedObjectsByType:
            for trackedObject in self.trackedObjectsByType[key]:
                resultArray.append(deepcopy(trackedObject))
        return resultArray

    def getNorm(self, vector):
        result = 0
        for i in vector:
            result = result + vector[i]*vector[i]
        return result


    def predictPhase(self,time):
        for key in self.trackedObjectsByType:
            for i in range(0,len(self.trackedObjectsByType[key])):

                time_since_last_predict = time - self.trackedObjectsByType[key][i].last_predict

                if time_since_last_predict > 0:

                    state = np.array([self.trackedObjectsByType[key][i].position[0],self.trackedObjectsByType[key][i].position[1],
                            self.trackedObjectsByType[key][i].velocity[0]*time_since_last_predict,
                            self.trackedObjectsByType[key][i].velocity[1]*time_since_last_predict])


                    estimate, covariance_sqrt = ddekf.predictPhase(self.predictFunction, self.predictFunctionJacobian,time,
                        self.trackedObjectsByType[key][i].velocity_covariance_sqrt,state,self.processNoiseSqrt)


                    self.trackedObjectsByType[key][i].velocity = estimate[2:4]*(1.0/time_since_last_predict)
                    self.trackedObjectsByType[key][i].velocity_covariance_sqrt = covariance_sqrt
                    self.trackedObjectsByType[key][i].position = estimate[0:2]
                    self.trackedObjectsByType[key][i].last_predict = time


    def updatePhase(self,measurements,time):
        measurementByType = {}
        for measurement in measurements:
            if measurement.objectType not in measurementByType:
                measurementByType[measurement.objectType] = []
            measurementByType[measurement.objectType].append(measurement)

        for key in measurementByType:
            if key not in self.trackedObjectsByType:
                self.trackedObjectsByType[key] = []
            self.updatePhaseByObject(measurementByType[key],time,key)


        #identify objects that are too OLD
        for key in self.trackedObjectsByType:
            indexToPop = []
            for i in range(0,len(self.trackedObjectsByType[key])):
                if self.threshold < ( time - self.trackedObjectsByType[key][i].last_update ):
                    indexToPop.append(i)
            for index in sorted(indexToPop,reverse=True):
                del self.trackedObjectsByType[key][index]

    def updatePhaseByObject(self,measurements,time,key):
        
        for i in range(0,len(measurements)):
            if len(self.trackedObjectsByType[key]) == 0:
                self.candidateCount = self.candidateCount + 1
                self.trackedObjectsByType[key].append( 
                    trackedObject(measurements[i].position,measurements[i].width,
                    measurements[i].height,self.initialVelocity,
                    self.initialCovariance,time,"id-"+str(self.candidateCount),measurements[i].objectType))
                continue
            IOUTrackedObjectsToMeasurement = []
            for k in range(0,len(self.trackedObjectsByType[key])):
                iou = intersection.IOU(measurements[i].x,measurements[i].y,measurements[i].width,measurements[i].height,
                    self.trackedObjectsByType[key][k].position[0],self.trackedObjectsByType[key][k].position[1],
                    self.trackedObjectsByType[key][k].width,self.trackedObjectsByType[key][k].width);
                IOUTrackedObjectsToMeasurement.append(iou) 
            if max(IOUTrackedObjectsToMeasurement) < self.minimum_IOU: 
                self.candidateCount = self.candidateCount + 1
                self.trackedObjectsByType[key].append( 
                    trackedObject(measurements[i].position,measurements[i].width,
                    measurements[i].height,self.initialVelocity,
                    self.initialCovariance,time,"id-"+str(self.candidateCount),measurements[i].objectType))

        #more measurements than objects:
        #it means that there are two objects are really close together
        local_minimum_IOU = self.minimum_IOU
        while True:
            if local_minimum_IOU > 1:
                break
            if len(measurements) > len(self.trackedObjectsByType[key]):
                local_minimum_IOU = local_minimum_IOU + 0.1
                for i in range(0,len(measurements)):
                    IOUTrackedObjectsToMeasurement = []
                    for k in range(0,len(self.trackedObjectsByType[key])):
                        iou = intersection.IOU(measurements[i].x,measurements[i].y,measurements[i].width,measurements[i].height,
                            self.trackedObjectsByType[key][k].position[0],self.trackedObjectsByType[key][k].position[1],
                            self.trackedObjectsByType[key][k].width,self.trackedObjectsByType[key][k].width);
                        IOUTrackedObjectsToMeasurement.append(iou) 
                    if max(IOUTrackedObjectsToMeasurement) < local_minimum_IOU:
                        self.candidateCount = self.candidateCount + 1
                        self.trackedObjectsByType[key].append( 
                            trackedObject(measurements[i].position,measurements[i].width,
                            measurements[i].height,self.initialVelocity,
                            self.initialCovariance,time,"id-"+str(self.candidateCount),measurements[i].objectType))
            else:
                break

        #compute IOU
        candidateCount = len(self.trackedObjectsByType[key])
        candidateIDs = []
        for k in range(0,len(self.trackedObjectsByType[key])):
            candidateIDs.append(self.trackedObjectsByType[key][k])

        candidateOptions = [] 
        maxCandidate = 0
        for k in range(0,len(self.trackedObjectsByType[key])):
            optionsForKthObject = []
            for i in range(0,len(measurements)):
                iou = intersection.IOU(measurements[i].x,measurements[i].y,measurements[i].width,measurements[i].height,
                    self.trackedObjectsByType[key][k].position[0][0],self.trackedObjectsByType[key][k].position[1][0],
                    self.trackedObjectsByType[key][k].width,self.trackedObjectsByType[key][k].width)
                optionsForKthObject.append(iou)
                maxCandidate = max(maxCandidate,iou)
            candidateOptions.append(optionsForKthObject)

        #append dummy objects
        dummyCount = 0
        if len(measurements) > len(self.trackedObjectsByType[key]):
            dummyCount = len(measurements) - len(self.trackedObjectsByType[key])
            for i in range(0,dummyCount):
                candidateOptions.append([0]*len(measurements))
                candidateIDs.append("dummy"+str(i))
        candidateCount = candidateCount + dummyCount
        
        for k in range(0,len(self.trackedObjectsByType[key])):
            for i in range(0,len(measurements)):
                candidateOptions[k][i] = maxCandidate - candidateOptions[k][i]
        
        extraCount = 0
        if candidateCount > len(measurements):
            extraCount = candidateCount - len(measurements)
            for row in range(0,candidateCount):
                for i in range(0,extraCount):
                    candidateOptions[row].append(maxCandidate+1)

        selection = hungarian.hungarianAlgorithm(candidateCount,candidateIDs,candidateOptions)

        if selection == None:
            return

        #after matching boxes using hungarian algorithm
        #only matching stuff as some trackedObjects might be missing measurements
        for i in range(0,len(selection) - dummyCount):
            if (selection[i][1] >= len(measurements)):#other case
                continue
            
            #if the matched measurement's objectType does not match with the trackedObject's objectType
            #we ignore this matching
            if (measurements[selection[i][1]].objectType != self.trackedObjectsByType[key][i].objectType):
                continue
        
            state = np.array([self.trackedObjectsByType[key][i].position[0],self.trackedObjectsByType[key][i].position[1],
                    self.trackedObjectsByType[key][i].velocity[0],self.trackedObjectsByType[key][i].velocity[1]])
            measurement = np.array([measurements[selection[i][1]].position[0],measurements[selection[i][1]].position[1]])

            estimate,covariance_sqrt = ddekf.updatePhase( self.sensorNoiseSqrt, self.trackedObjectsByType[key][i].velocity_covariance_sqrt,
                self.measurementMatrix,state,measurement) 

            self.trackedObjectsByType[key][i].velocity = estimate[2:4]
            self.trackedObjectsByType[key][i].position = estimate[0:2]
            self.trackedObjectsByType[key][i].velocity_covariance_sqrt = covariance_sqrt

            self.trackedObjectsByType[key][i].width = int(float(self.trackedObjectsByType[key][i].width)*(1.0 - self.newShapeWeight) +  float(measurements[selection[i][1]].width)*(self.newShapeWeight))
            self.trackedObjectsByType[key][i].height = int(float(self.trackedObjectsByType[key][i].height)*(1.0 - self.newShapeWeight) +  float(measurements[selection[i][1]].height)*(self.newShapeWeight))
            self.trackedObjectsByType[key][i].last_update = time
            self.trackedObjectsByType[key][i].last_predict = time

File: java_implementation/Sort.java
import java.util.ArrayList;
import java.util.*; 
import java.util.Collections;

class Sort{
	Map<String,ArrayList<TrackedObject>> trackedObjectsByType;
	Matrix processNoise;
	Matrix sensorNoise;
	Matrix processNoiseSqrt;
	Matrix sensorNoiseSqrt;
	Matrix measurementMatrix;
	double minimum_IOU;
	double threshold;
	int totalCandidateCount;
	double newShapeWeight;
	Function predictFunction;
	Function predictFunctionDefault = new FunctionLinear();
	Matrix initialCovariance;
	Matrix initialVelocity;

	public Sort(Matrix sensorNoise, Matrix processNoise, Function f, double minimumIOU, double newShapeWeight, double threshold, Matrix initialVelocity, Matrix initialCovariance){
		this.trackedObjectsByType = new HashMap<String,ArrayList<TrackedObject> >();
		this.processNoise = new Matrix(processNoise);
		this.sensorNoise = new Matrix(sensorNoise);
		this.processNoiseSqrt = new Matrix(processNoise);
		this.sensorNoiseSqrt = new Matrix(sensorNoise);
		this.measurementMatrix = Matrix.zero(2,4);
		this.measurementMatrix.set(0,0,1);
		this.measurementMatrix.set(1,1,1);
		this.minimum_IOU = minimumIOU;
		this.threshold = threshold;
		this.totalCandidateCount = 0;
		this.newShapeWeight = newShapeWeight;

		if(f==null){
			this.predictFunction = this.predictFunctionDefault;
		}
		else{
			this.predictFunction =  f;
		}

		this.initialCovariance = new Matrix(initialCovariance);
		this.initialVelocity = new Matrix(initialVelocity);
	}

	public void set_minimumIOU(double IOU){
		this.minimum_IOU = IOU;
	}

	public void set_threshold(double threshold){
		this.threshold = threshold;
	}

	public void set_sensorNoise(Matrix sensorNoise){
		this.sensorNoise = new Matrix(sensorNoise);
		this.sensorNoiseSqrt = new Matrix(sensorNoise);
	}

	public Matrix get_sensorNoise(){
		return new Matrix(this.sensorNoise);
	}

	public void set_processNoise(Matrix processNoise){
		this.processNoise = new Matrix(processNoise);
		this.processNoiseSqrt = new Matrix(processNoise);
	}

	public Matrix get_processNoise(){
		return new Matrix(this.processNoise);
	}

	public void set_predictFunction(Function f){
		this.predictFunction = f;
	}

	public void set_newShapeWeight(double weight){
		this.newShapeWeight = weight;
	}


	public TrackedObject[] getTracked(){
		int totalCount = 0;
		Iterator <String> it = trackedObjectsByType.keySet().iterator();	
		while(it.hasNext()){
			String key = it.next();
			totalCount += trackedObjectsByType.get(key).size();

		}
		TrackedObject[] result = new TrackedObject[totalCount];

		it = trackedObjectsByType.keySet().iterator();	
		int counter = 0;
		while(it.hasNext()){
			String key = it.next();
			for(int i=0; i<this.trackedObjectsByType.get(key).size(); i++){
				result[counter++] = this.trackedObjectsByType.get(key).get(i);
			}
		}
		return result;
	}


	public void predictPhase(double time){
		Iterator <String> it = trackedObjectsByType.keySet().iterator();	
		while(it.hasNext()){
			String key = it.next();
			for(int i=0; i<trackedObjectsByType.get(key).size(); i++){
				
				double time_since_last_predict = time - trackedObjectsByType.get(key).get(i).last_predict;

				if(time_since_last_predict > 0){
					Matrix state = new Matrix(4,1); 
					state.set(0,0,trackedObjectsByType.get(key).get(i).position.get(0,0));
					state.set(1,0,trackedObjectsByType.get(key).get(i).position.get(1,0));
					state.set(2,0,trackedObjectsByType.get(key).get(i).velocity.get(0,0)*time_since_last_predict);
					state.set(3,0,trackedObjectsByType.get(key).get(i).velocity.get(1,0)*time_since_last_predict);

					Matrix[] results = KalmanFilter.predictPhase( this.predictFunction, time, this.trackedObjectsByType.get(key).get(i).velocity_covariance_sqrt, state, this.processNoiseSqrt );

					Matrix estimate = results[0];
					Matrix covariance_sqrt = results[1];



					this.trackedObjectsByType.get(key).get(i).velocity = Matrix.scalarMultiply(1.0/(time_since_last_predict),estimate.getSubmatrix(2,0,2,1));
					this.trackedObjectsByType.get(key).get(i).velocity_covariance_sqrt = covariance_sqrt;
					this.trackedObjectsByType.get(key).get(i).position = estimate.getSubmatrix(0,0,2,1);
					this.trackedObjectsByType.get(key).get(i).last_predict = time;
				}
			}
		}
	}

	public void updatePhase( Measurement[] measurements, double time ){
		Map<String,ArrayList<Measurement>> measurementsByType = new HashMap<String,ArrayList<Measurement>>();
		for(int i=0; i<measurements.length; i++){
			if(!measurementsByType.containsKey( measurements[i].objectType )){
				measurementsByType.put(measurements[i].objectType, new ArrayList<Measurement>());
			}
			measurementsByType.get(measurements[i].objectType).add(measurements[i]);
		}

		Iterator <String>it = measurementsByType.keySet().iterator();	
		while(it.hasNext()){
			String key = it.next();
			if(!this.trackedObjectsByType.containsKey(key)){
				this.trackedObjectsByType.put(key, new ArrayList<TrackedObject>());
			}
			this.updatePhaseByObject(measurementsByType.get(key),time,key);
		}

		//find objects that are too old
		it = this.trackedObjectsByType.keySet().iterator();	
		while(it.hasNext()){
			String key = it.next();
			ArrayList<Integer> indexToPop = new ArrayList<Integer>();
			for(int i=0; i<this.trackedObjectsByType.get(key).size(); i++){
				if(this.threshold < (time-this.trackedObjectsByType.get(key).get(i).last_update)){
					indexToPop.add(i);
				}
			}
			Collections.sort(indexToPop);
			int deletedSoFar = 0;
			for(int i=0; i<indexToPop.size(); i++){
				this.trackedObjectsByType.get(key).remove(indexToPop.get(i)-deletedSoFar);
				deletedSoFar++;
			}
		}
	}

	public void updatePhaseByObject(ArrayList<Measurement> measurements, double time, String key){
		//identify new objects
		for(int i=0; i<measurements.size(); i++){
			if(this.trackedObjectsByType.get(key).size() == 0){
				this.totalCandidateCount = this.totalCandidateCount + 1;
				this.trackedObjectsByType.get(key).add(new TrackedObject(measurements.get(i).position,measurements.get(i).width,measurements.get(i).height,this.initialVelocity,this.initialCovariance,time,"id-"+this.totalCandidateCount,measurements.get(i).objectType));
				continue;
			}
			ArrayList<Double> IOUTrackedObjectsToMeasurement = new ArrayList<Double>();
			double maxIOU = 0;
			for( int k=0; k<this.trackedObjectsByType.get(key).size(); k++){
				double iou = Intersection.IOU(measurements.get(i).x,measurements.get(i).y,measurements.get(i).width,measurements.get(i).height,
					this.trackedObjectsByType.get(key).get(k).position.get(0,0),this.trackedObjectsByType.get(key).get(k).position.get(1,0),
					this.trackedObjectsByType.get(key).get(k).width,this.trackedObjectsByType.get(key).get(k).height);
				IOUTrackedObjectsToMeasurement.add(iou);
				maxIOU = Math.max(maxIOU,iou);
			}
			if(maxIOU < this.minimum_IOU){
				this.totalCandidateCount = this.totalCandidateCount + 1;
				this.trackedObjectsByType.get(key).add(new TrackedObject(measurements.get(i).position,measurements.get(i).width,measurements.get(i).height,this.initialVelocity,this.initialCovariance,time,"id-"+this.totalCandidateCount,measurements.get(i).objectType));
			}
		}

		double local_minimum_IOU = this.minimum_IOU;
		while(true){
			if(local_minimum_IOU > 1){
				break;
			}
			if( measurements.size() > this.trackedObjectsByType.get(key).size()){
				local_minimum_IOU += 0.1;
				for(int i=0; i<measurements.size(); i++){
					ArrayList<Double> IOUTrackedObjectsToMeasurement = new ArrayList<Double>();
					double maxIOU = 0;
					for(int k=0; k<this.trackedObjectsByType.get(key).size(); k++){
						double iou = Intersection.IOU(measurements.get(i).x,measurements.get(i).y,measurements.get(i).width,measurements.get(i).height,
							this.trackedObjectsByType.get(key).get(k).position.get(0,0),this.trackedObjectsByType.get(key).get(k).position.get(1,0),
							this.trackedObjectsByType.get(key).get(k).width,this.trackedObjectsByType.get(key).get(k).height);
						IOUTrackedObjectsToMeasurement.add(iou);
						maxIOU =Math.max(maxIOU,iou);
					}
					if(maxIOU < this.minimum_IOU){
						this.totalCandidateCount = this.totalCandidateCount + 1;
						this.trackedObjectsByType.get(key).add(new TrackedObject(measurements.get(i).position,measurements.get(i).width,measurements.get(i).height,this.initialVelocity,this.initialCovariance,time,"id-"+this.totalCandidateCount,measurements.get(i).objectType));
					}
				}
			}
			else{
				break;
			}
		}

		int localCandidateCount = this.trackedObjectsByType.get(key).size();
		ArrayList<String> candidateIDs = new ArrayList<String>();

		for(int k=0; k<this.trackedObjectsByType.get(key).size(); k++){
			candidateIDs.add(this.trackedObjectsByType.get(key).get(k).name); 
		}

		ArrayList<ArrayList<Double>> candidateOptions = new ArrayList<ArrayList<Double>>();
		double maxCandidate = 0;
		for(int k=0; k<this.trackedObjectsByType.get(key).size(); k++){
			ArrayList<Double> optionsForKthObject = new ArrayList<Double>();
			for(int i=0; i<measurements.size(); i++){
				double iou = Intersection.IOU(measurements.get(i).x,measurements.get(i).y,measurements.get(i).width,measurements.get(i).height,
					this.trackedObjectsByType.get(key).get(k).position.get(0,0),this.trackedObjectsByType.get(key).get(k).position.get(1,0),
					this.trackedObjectsByType.get(key).get(k).width,this.trackedObjectsByType.get(key).get(k).height);
				optionsForKthObject.add(iou);
				maxCandidate = Math.max(maxCandidate,iou);
			}
			candidateOptions.add(optionsForKthObject);
		}

		int dummyCount = 0;
		if(measurements.size() > this.trackedObjectsByType.get(key).size()){
			dummyCount = measurements.size() - this.trackedObjectsByType.get(key).size();
			for(int i=0; i<dummyCount; i++){
				ArrayList<Double> optionsForDummyObject = new ArrayList<Double>();
				for(int k=0; k<measurements.size(); k++){
					optionsForDummyObject.add(0.0);
				}
				candidateOptions.add(optionsForDummyObject);
				candidateIDs.add("dummy"+i);
			}
		}
		
		localCandidateCount += dummyCount;

		
		for(int k=0; k<this.trackedObjectsByType.get(key).size(); k++){
			for(int i=0; i<measurements.size(); i++){
				candidateOptions.get(k).set(i,maxCandidate - candidateOptions.get(k).get(i));
			}
		}


		int extraCount = 0;
		if(localCandidateCount > measurements.size()){
			extraCount = localCandidateCount - measurements.size();
			for(int row=0; row<localCandidateCount; row++){
				for(int i=0; i<extraCount; i++){
					candidateOptions.get(row).add(maxCandidate+1);
				}
			}
		}

		double [][] options = new double[localCandidateCount][localCandidateCount];
		String [] IDs = new String[localCandidateCount];
		for(int i=0; i<localCandidateCount; i++){
			for(int j=0; j<localCandidateCount; j++){
				options[i][j] = candidateOptions.get(i).get(j);
			}
			IDs[i] = candidateIDs.get(i);
		}

		hungarian solver = new hungarian();
		List<hungarian.Selection> selection = solver.hungarianAlgorithm(localCandidateCount,IDs,options);
		if(selection == null){
			return;
		}

		for(int i=0; i<selection.size()-dummyCount;i++){
			if(selection.get(i).jobNum >= measurements.size()){
				continue;
			}


			Matrix state = Matrix.zero(4,1);
			state.setSubmatrix(0,0,2,1,this.trackedObjectsByType.get(key).get(i).position);
			state.setSubmatrix(2,0,2,1,this.trackedObjectsByType.get(key).get(i).velocity);
			Matrix measurement = new Matrix(measurements.get(selection.get(i).jobNum).position);

			Matrix [] result = KalmanFilter.updatePhase( this.sensorNoiseSqrt, this.trackedObjectsByType.get(key).get(i).velocity_covariance_sqrt,
				this.measurementMatrix,state,measurement);

			Matrix estimate = new Matrix(result[0]);
			Matrix covariance_sqrt = new Matrix(result[1]);

			this.trackedObjectsByType.get(key).get(i).position = estimate.getSubmatrix(0,0,2,1);
			this.trackedObjectsByType.get(key).get(i).velocity = estimate.getSubmatrix(2,0,2,1);
			this.trackedObjectsByType.get(key).get(i).velocity_covariance_sqrt = new Matrix(covariance_sqrt);

			this.trackedObjectsByType.get(key).get(i).width = (this.trackedObjectsByType.get(key).get(i).width)*(1.0-this.newShapeWeight) + (measurements.get(selection.get(i).jobNum).width*this.newShapeWeight);
			this.trackedObjectsByType.get(key).get(i).height = (this.trackedObjectsByType.get(key).get(i).height)*(1.0-this.newShapeWeight) + (measurements.get(selection.get(i).jobNum).height*this.newShapeWeight);
			this.trackedObjectsByType.get(key).get(i).last_update = time;
			this.trackedObjectsByType.get(key).get(i).last_predict = time;
		}
	}
}
